lib/auth/accountrecovery.go

```diff
@@ -33,13 +37,113 @@ import (
+	}
+
+	if err := s.verifyCodeWithRecoveryLock(ctx, req.GetUsername(), req.GetRecoveryCode()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Remove any other existing tokens for this user before creating a token.
+	if err := s.deleteUserTokens(ctx, req.Username); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	token, err := s.createRecoveryToken(ctx, req.GetUsername(), UserTokenTypeRecoveryStart, req.GetRecoverType())
+	return token, trace.Wrap(err)
+}
+
+// verifyCodeWithRecoveryLock counts number of failed attempts at providing a valid recovery code.
+// After MaxAccountRecoveryAttempts, user is temporarily locked from further attempts at recovering and also
+// locked from logging in. Modeled after existing function WithUserLock.
+func (s *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
+	user, err := s.Identity.GetUser(username, false)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			// If user is not found, still authenticate. It should
+			// always return an error. This prevents username oracles and
+			// timing attacks.
+			return s.verifyRecoveryCode(ctx, username, recoveryCode)
+		}
+		return trace.Wrap(err)
+	}
+
+	status := user.GetStatus()
+	if status.IsLocked && status.RecoveryAttemptLockExpires.After(s.clock.Now().UTC()) {
+		log.Debugf("%v exceeds %v failed account recovery attempts, locked until %v",
+			user.GetName(), defaults.MaxAccountRecoveryAttempts, apiutils.HumanTimeFormat(status.RecoveryAttemptLockExpires))
+		return trace.AccessDenied("too many incorrect recovery attempts, please try again later")
+	}
+
+	fnErr := s.verifyRecoveryCode(ctx, username, recoveryCode)
+	if fnErr == nil {
+		return nil
+	}
+
+	// Do not lock user in case if DB is flaky or down.
+	if trace.IsConnectionProblem(fnErr) {
+		return trace.Wrap(fnErr)
+	}
+
+	// Log failed attempt.
+	now := s.clock.Now().UTC()
+	attempt := &types.RecoveryAttempt{Time: now, Expires: now.Add(defaults.AttemptTTL)}
+	if err := s.CreateUserRecoveryAttempt(ctx, username, attempt); err != nil {
+		log.Error(trace.DebugReport(err))
+		return trace.Wrap(fnErr)
+	}
+
+	attempts, err := s.Identity.GetUserRecoveryAttempts(ctx, username)
+	if err != nil {
+		log.Error(trace.DebugReport(err))
+		return trace.Wrap(fnErr)
+	}
+
+	if !types.IsMaxFailedRecoveryAttempt(defaults.MaxAccountRecoveryAttempts, attempts, now) {
+		log.Debugf("%v user has less than %v failed account recovery attempts", username, defaults.MaxAccountRecoveryAttempts)
+		return trace.Wrap(fnErr)
+	}
+
+	// Reached max attempts.
+	lockUntil := s.clock.Now().UTC().Add(defaults.AccountLockInterval)
+
+	log.Debugf("%v exceeds %v failed account recovery attempts, account locked until %v and an email has been sent",
+		username, defaults.MaxAccountRecoveryAttempts, apiutils.HumanTimeFormat(lockUntil))
+
+	// Temp lock both user login and recovery attempts.
+	user.SetLockedFromRecoveryAttempt(lockUntil)
+	user.SetLocked(lockUntil, accountLockedMsg)
```

Comment:
Are those calls always a pair? If yes, why not do `user.SetLockedFromRecoveryAttempt(lockUntil, accountLockedMsg)` and have that method call SetLocked for us?